### PR TITLE
Log warnings that are on the second level also

### PR DESCRIPTION
**Pull request by Albert221**
Bug: T177191

/cc @samwilson

### CODE REVIEW

File: src/MediawikiApi.php [hunk 1]
```diff
@@ -351,14 +351,36 @@ private function getUserAgent() {
...
+			}
+
+			// ...if no then go one level deeper and check there for them.
+			foreach ( $result as $value ) {
+				if ( !array_key_exists( 'warnings', $value ) ) {
+					return;
+				}
+
+				foreach ( $value['warnings'] as $module => $warningData ) {
+					// Accomodate both formatversion=2 and old-style API results
```
Comment (samwilson) [inline src/MediawikiApi.php]: This section should not be duplicated from above. Instead, you could a new protected method `logWarning()` and call it from each of these places.

File: src/MediawikiApi.php [hunk 2]
```diff
@@ -351,14 +351,36 @@ private function getUserAgent() {
...
+				foreach ( $result['warnings'] as $module => $warningData ) {
+					// Accomodate both formatversion=2 and old-style API results
+					$logPrefix = $module . ': ';
+					if ( isset( $warningData['*'] ) ) {
+						$this->logger->warning( $logPrefix . $warningData['*'], [ 'data' => $warningData ] );
+					} else {
+						$this->logger->warning( $logPrefix . $warningData['warnings'], [ 'data' => $warningData ] );
+					}
+
+					return;
```
Comment (samwilson) [inline src/MediawikiApi.php]: You're returning after the first iteration of the loop. What about if there are multiple warnings?

File: src/MediawikiApi.php [hunk 3]
```diff
@@ -351,14 +351,36 @@ private function getUserAgent() {
...
+				}
+			}
+
+			// ...if no then go one level deeper and check there for them.
+			foreach ( $result as $value ) {
+				if ( !array_key_exists( 'warnings', $value ) ) {
```
Comment (samwilson) [inline src/MediawikiApi.php]: `$value` can be a string (see [test results](https://travis-ci.org/addwiki/mediawiki-api-base/jobs/318353872#L999)).

**Review by samwilson — CHANGES_REQUESTED** [review]
Thanks for working on this. A couple of comments.

Note that you could also add a new test, perhaps with the duplicate-upload warning as given in the ticket; that'd make sure this was fixed correctly. :-)

File: src/MediawikiApi.php [hunk 4]
```diff
@@ -351,19 +351,52 @@ private function getUserAgent() {
...
 			}
 		}
 	}
 
+	/**
+	 * @param $array
+	 *
+	 * @return bool Whether any warning has been logged or not.
+	 */
+	private function logWarning( $array ) {
```
Comment (samwilson) [inline src/MediawikiApi.php]: This should be proteccted, in case subclasses want to do something different with it.

**Review by samwilson — CHANGES_REQUESTED** [review]
Looking good!

Just that little coding-standard fix that Travis is complaining about, and one minor comment.

**Review by samwilson — APPROVED** [review]
Looks good to me. @addshore are you okay with this?

**Review by addshore — APPROVED** [review]
